File: NeetCode/1-D_Dynamic_Programming/Coin_Change/coin__change.py
from typing import Optional, List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float("inf") for _ in range(amount + 1)]
        dp[0] = 0

        for amount in range(1, amount + 1):
            for c in coins:
                if amount - c >= 0:
                    dp[amount] = min(dp[amount], dp[amount - c] + 1)
        return dp[amount] if dp[amount] != float("inf") else -1


if __name__ == "__main__":
    sol = Solution()
    coins = [1, 5, 10, 100, 200]
    amount = 12
    print(sol.coinChange(coins, amount))
    print("Running Solution...")

File: NeetCode/1-D_Dynamic_Programming/Coin_Change/Coin_Change.cpp
// Coin_Change.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        vector<long long> dp(amount + 1, INT_MAX);
        dp[0] = 0;

        for (int amt = 1; amt < amount + 1; amt++)
        {
            for (auto coin : coins)
            {
                if (amt - coin >= 0)
                {
                    dp[amt] = min(dp[amt], dp[amt - coin] + 1);
                }
            }
        }
        if (dp[amount] == INT32_MAX)
        {
            return -1;
        }
        return dp[amount];
    }
};
